burn-wgpu/src/template/matmul_blocktiling_2d.wgsl

```diff
@@ -0,0 +1,133 @@
+@group(0)
+@binding(0)
+var<storage, read> lhs: array<{{ elem }}>;
+
+@group(0)
+@binding(1)
+var<storage, read> rhs: array<{{ elem }}>;
+
+@group(0)
+@binding(2)
+var<storage, read_write> output: array<{{ elem }}>;
+
+@group(0)
+@binding(3)
+var<storage, read> info: array<u32>;
+
+const B_M = {{b_m}}u;
+const B_N = {{b_n}}u;
+const B_K = {{b_k}}u;
+const B_M_X_B_K = {{bm_x_bk}}u;
+const B_K_X_B_N = {{bk_x_bn}}u;
+const T_M = {{t_m}}u;
+const T_N = {{t_n}}u;
+const T_M_X_T_N = {{tm_x_tn}}u;
+
+var<workgroup> shared_lhs: array<{{ elem }}, B_M_X_B_K>; 
+var<workgroup> shared_rhs: array<{{ elem }}, B_K_X_B_N>;
+
+@compute
+@workgroup_size({{ workgroup_size_x }}, {{ workgroup_size_y }}, {{ workgroup_size_z }})
+fn main(
+    @builtin(global_invocation_id) global_id: vec3<u32>,
+    @builtin(local_invocation_index) local_idx: u32,
+    @builtin(workgroup_id) workgroup_id: vec3<u32>,
+) {
+    let skip_row = workgroup_id.x * B_M;
+    let skip_col = workgroup_id.y * B_N;
+
+    let n_thread_per_row = ((B_N - 1u) / T_N) + 1u;
+    let thread_row = (local_idx / n_thread_per_row) * T_M;
+    let thread_col = (local_idx % n_thread_per_row) * T_N;
+    
+    let row = skip_row + thread_row;
+    let col = skip_col + thread_col;
+
+    let batch = global_id.z;
+
+    // Basic information
+    let dim = info[0];
+    let n_rows = info[6u * dim - 1u]; 
+    let n_cols = info[6u * dim]; 
+    let K = info[5u * dim - 1u];
+
+    // Calculate the corresponding offsets with support for broadcasting.
+    let offset_output = batch * n_rows * n_cols; 
+    var offset_lhs: u32 = skip_row * K; 
+    var offset_rhs: u32 = skip_col;
+
+    let batch_dims = dim - 2u;
+    for (var b: u32 = 1u; b <= batch_dims; b++) {
+        let stride_lhs = info[b];
+        let stride_rhs = info[b + dim];
+        let stride_output = info[b + 2u * dim];
+        let shape_lhs = info[b + 3u * dim];
+        let shape_rhs = info[b + 4u * dim];
+
+        offset_lhs += offset_output / stride_output % shape_lhs * stride_lhs;
+        offset_rhs += offset_output / stride_output % shape_rhs * stride_rhs;
+    }
+    
+    // in case T_ does not divide B_ evenly
+    let actual_T_M = min(B_M - thread_row, T_M);
+    let actual_T_N = min(B_N - thread_col, T_N);
+
+    var results: array<{{ elem }}, T_M_X_T_N>;
+    var register_M: array<{{ elem }}, T_M>;
+    var register_N: array<{{ elem }}, T_N>;
+
+    for (var k = 0u; k < K; k += B_K) { 
+        let sm_limit = min(B_K, K - k);
+        
```

Comment:
For big blocks of code, can we add a brief intention comment? It's a good practice in general.